pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/ModelParser.groovy

```diff
@@ -271,6 +283,71 @@ class ModelParser implements Parser {
         return r;
     }
 
+    /**
+     * This works functionally equivalent to the string used in {@link #parseArgument}, but without returning a
+     * {@link ModelASTValue}, returning just a string instead, and without handling {@link MapExpression} or
+     * {@link VariableExpression}, which is exclusively used for the `agent any` and `agent none` shortcuts regardless.
+     *
+     * @param e A non-null expression. If the expression is anything but a constant or a GString, an error will be thrown.
+     * @return The string value for that expression - in the case of {@link GStringExpression}, we explicitly remove the
+     * quotes from around it so that it's suitable for string concatenation. Returns null if an error was encountered.
+     */
+    @CheckForNull
+    private String envStringFromArbitraryExpression(@Nonnull Expression e) {
+        StringBuilder builder = new StringBuilder()
+        if (e instanceof ConstantExpression) {
+            builder.append(e.value)
+        } else if (e instanceof GStringExpression) {
+            String rawSrc = getSourceText(e)
```

Comment:
Nope! `FOO = "A" + "${return 'B'}"` will end up equivalent to `FOO = "A${return 'B'}"`. The actual `GString` will still get evaluated at the same time no matter which way it's specified.